py/server/deephaven/experimental/partitioned_table_service.py

```diff
@@ -0,0 +1,326 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+
+from abc import ABC, abstractmethod
+from typing import Tuple, Optional, Any, Callable
+
+import jpy
+
+import pyarrow as pa
+
+from deephaven.dherror import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.table import Table
+
+_JPythonTableDataService = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService")
+_JTableKeyImpl = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService$TableKeyImpl")
+_JTableLocationKeyImpl = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService$TableLocationKeyImpl")
+
+
+class TableKey:
+    """A key that identifies a table. The key should be unique for each table. The key can be any Python object and
+    should include sufficient information to uniquely identify the table for the backend service."""
+
+    def __init__(self, key: Any):
+        self._key = key
+
+    @property
+    def key(self) -> Any:
+        """The user defined key that identifies the table."""
+        return self._key
+
+
+class PartitionedTableLocationKey:
+    """A key that identifies a specific partition of a table. The key should be unique for each partition of the table.
+    The key can be any Python object and should include sufficient information to uniquely identify the partition for
+    the backend service to fetch the partition data.
+    """
+
+    def __init__(self, pt_location_key: Any):
+        self._pt_location_key = pt_location_key
+
+    @property
+    def pt_location_key(self) -> Any:
+        return self._pt_location_key
+
+
+
+class PartitionedTableServiceBackend(ABC):
+    """An interface for a backend service that provides access to partitioned data."""
+
+    @abstractmethod
+    def table_schema(self, table_key: TableKey) -> Tuple[pa.Schema, Optional[pa.Schema]]:
+        """ Returns the table schema and optionally the schema for the partition columns for the table with the given
+        table key.
+        The table schema is not required to include the partition columns defined in the partition schema. The
+        partition columns are limited to primitive types and strings.
+
+        Args:
+            table_key (TableKey): the table key
+
+        Returns:
+            Tuple[pa.Schema, Optional[pa.Schema]]: a tuple of the table schema and the optional schema for the partition
+                columns
```

Comment:
calling this stuff "optional schema" is very confusing.  It makes it sound like the user has the option to return the value, but they clearly do not.

Reply:
What would it look like to make it clear that it is optional?

When this is not supplied, the user is communicating that there are no partitioning columns. They could also send an empty table; but we can do that for them making it very clear that it is optional. (At least; this is my understanding.)

Reply:
The key detail is that if it is not supplied the user is communicating that there are no partitioning columns.  This is not made clear in the docsstring.